### PR TITLE
net: initial import of an IPv6 neighbor cache

### PR DESCRIPTION
**Pull request by authmillenon**
~~Depends on #2433~~ merged
~~Depends on #2546 for debug output~~ merged

### CODE REVIEW

File: sys/include/net/ng_ipv6.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2014 Martin Lenders <mlenders@inf.fu-berlin.de>
```
Comment (Lotterleben) [inline sys/include/net/ng_ipv6.h]: Aren't this date & name deprecated? :)

Comment (authmillenon) [inline sys/include/net/ng_ipv6.h]: Yes. It seems I have a broken template somewhere :D

Comment (authmillenon) [inline sys/include/net/ng_ipv6.h]: (This file is part of #2433 but thanks, never the less)

File: sys/include/net/ng_ipv6/nc.h [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+/**
+ * @}
+ */
+/**
+ * @}
+ */
+
+/**
+ * @brief   Neighbor cache entry as defined in
+ *          <a href="http://tools.ietf.org/html/rfc4861#section-3.5#section-5.1">
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/nc.h]: `s/#section-3.5//`

File: sys/include/net/ng_ipv6/nc.h [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+ *              RFC 4861, section 5.1
+ *          </a>.
+ */
+typedef struct ng_ipv6_nc {
+    struct ng_ipv6_nc *next;    /**< Next entry for this interface */
+    pktqueue_t pkts;            /**< Packets waiting for address resolution */
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/nc.h]: Why is this required?

File: sys/include/net/ng_ipv6/nc.h [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    ng_ipv6_addr_t ipv6_addr;   /**< IPv6 address of the neighbor */
+    uint8_t l2_addr[8];         /**< Link layer address of the neighbor */
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/nc.h]: Why 8?

File: sys/include/net/ng_ipv6/nc.h (hunk 3 above)
Comment (authmillenon) [inline sys/include/net/ng_ipv6/nc.h]: According to the further above referenced [RFC 4861](http://tools.ietf.org/html/rfc4861#section-5.1)

> Entries […] contain such information as […] a pointer to any queued packets waiting for address resolution to complete, etc

And after seeing this I thought this might be the easiest solution to trigger a send of the packet, once the address was resolved. Since you need the entry anyways (to add the address) you can just send all queued packets from here on.

Comment (OlegHahm) [inline sys/include/net/ng_ipv6/nc.h]: Sorry, apparently I skipped this. Seems practicable.

**Comment by authmillenon** [comment]
Rebased to current master

File: sys/include/net/ng_ipv6/nc.h [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+#include "timex.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   The size of the neighbor cache
+ */
+#define NG_IPV6_NC_SIZE             (NG_NETIF_NUMOF * 8)
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/nc.h]: is 8 big enough? I would definitely put an `#ifndef...` around this, so that the application designer - who might know more about expected node degree - can overwrite this.

File: sys/include/net/ng_ipv6/nc.h [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+/**
+ * @brief   The maximum size of a link layer address
+ */
+#define NG_IPV6_L2_ADDR_MAX         (8)
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/nc.h]: Again, it might make sense to give other modules the option to overrule this. Think of some weird radio device with huge L2 address that might want to overwrite this.

File: sys/include/net/ng_ipv6/nc.h [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+ *          RFC 4861, section 7.3.2
+ *      </a>
+ */
+#define NG_IPV6_NC_STATE_MASK           (0x07)
+#define NG_IPV6_NC_STATE_POS            (0)
+
+#define NG_IPV6_NC_STATE_UNMANAGED      (0x00)  /**< The entry is not manage by NDP */
+
+/**
+ * @brief The entry is unreachable
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/nc.h]: Doesn't this line belongs a couple of lines below?

**Comment by OlegHahm** [comment]
I guess it's too earlier for a review, right?

**Comment by authmillenon** [comment]
I think so too ;-)

File: sys/include/net/ng_ipv6/nc.h (hunk 7 above)
Comment (authmillenon) [inline sys/include/net/ng_ipv6/nc.h]: no, it documents the `NG_IPV6_NC_STATE_UNREACHABLE` macro, which is exactly below this comment block.

**Comment by authmillenon** [comment]
No longer WIP (except for tests)

File: sys/include/net/ng_ipv6/nc.h (hunk 7 above)
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/nc.h]: Never mind! Will open my eyes next time.

File: sys/net/network_layer/ng_ipv6/nc/ng_ipv6_nc.c [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+#include <string.h>
+
+#include "net/ng_ipv6/addr.h"
+#include "net/ng_ipv6/nc.h"
+
+static ng_ipv6_nc_t ncache[NG_IPV6_NC_SIZE];
+
+void ng_ipv6_nc_init(void)
+{
+    memset(ncache, 0, sizeof(ncache));
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/nc/ng_ipv6_nc.c]: Superfluous. Static arrays are initialized to 0 anyway.

File: sys/net/network_layer/ng_ipv6/nc/ng_ipv6_nc.c [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    for (int i = 0; i < NG_IPV6_NC_SIZE; i++) {
+        if (ng_ipv6_addr_is_unspecified(&(ncache[i].ipv6_addr))) {
+            ncache[i].iface = iface;
+
+            ng_pktqueue_init(&(ncache[i].pkts));
+            memcpy(&(ncache[i].ipv6_addr), ipv6_addr, sizeof(ng_ipv6_addr_t));
+
+            if (l2_addr != NULL && l2_addr_len > 0) {
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/nc/ng_ipv6_nc.c]: parens

File: sys/net/network_layer/ng_ipv6/nc/ng_ipv6_nc.c [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+ng_ipv6_nc_t *ng_ipv6_nc_get(kernel_pid_t iface, const ng_ipv6_addr_t *ipv6_addr)
+{
+    if (ipv6_addr == NULL) {
+        return NULL;
+    }
+
+    for (int i = 0; i < NG_IPV6_NC_SIZE; i++) {
+        if ((iface != KERNEL_PID_UNDEF || iface == ncache[i].iface) &&
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/nc/ng_ipv6_nc.c]: parens

File: sys/net/network_layer/ng_ipv6/nc/ng_ipv6_nc.c (hunk 8 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/nc/ng_ipv6_nc.c]: No it's not. For testing this makes sense for reseting.

**Comment by OlegHahm** [comment]
I'm fine with the current state.

**Comment by authmillenon** [comment]
No longer WIP

**Comment by authmillenon** [comment]
Depends on #2546 for debug output

**Comment by authmillenon** [comment]
Rebased to current #2546.

File: tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+static void test_ipv6_nc_remove__no_entry_pid(void)
+{
+    ng_ipv6_addr_t addr = { {
+            0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+            0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f
+        }
+    };
+
+    test_ipv6_nc_add__success();
+    ng_ipv6_nc_remove(TEST_UINT16 + 1, &addr);
```
Comment (OlegHahm) [inline tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c]: Why `+ 1`?

File: tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+        }
+    };
+    ng_ipv6_addr_t exp_addr = { {
+            0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+            0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f
+        }
+    };
+
+    test_ipv6_nc_add__success();
+    ng_ipv6_nc_remove(TEST_UINT16, &addr);
```
Comment (OlegHahm) [inline tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c]: removing an address that hasn't been add it or do I got something wrong? Maybe some documentation would be helpful - even for unittests.

Comment (authmillenon) [inline tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c]: It was added in `test_ipv6_nc_add__success()`. But I can write it up.

File: tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c (hunk 11 above)
Comment (authmillenon) [inline tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c]: To have another interface than the already registered one

**Comment by authmillenon** [comment]
Addressed comments.

File: tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c (hunk 12 above)
Comment (OlegHahm) [inline tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c]: Wasn't rather `exp_addr` added?

Comment (authmillenon) [inline tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c]: Ah yes… but that's the point of this test: removing an address, that is not there. See 9986aef
 for clarification

Comment (authmillenon) [inline tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c]: The name of the test case is thought to describe it

**Comment by OlegHahm** [comment]
rebase is required.

File: tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+    ng_ipv6_nc_t *entry = NULL;
+
+    test_ipv6_nc_add__success(); /* adds DEFAULT_TEST_IPV6_ADDR to DEFAULT_TEST_NETIF */
+
+    TEST_ASSERT_NOT_NULL((entry = ng_ipv6_nc_get(KERNEL_PID_UNDEF, &addr)));
+    TEST_ASSERT_EQUAL_INT(DEFAULT_TEST_NETIF, entry->iface);
+    TEST_ASSERT(ng_ipv6_addr_equal(&(entry->ipv6_addr), &addr));
+    TEST_ASSERT_EQUAL_STRING(TEST_STRING4, (char *)entry->l2_addr);
+    TEST_ASSERT_EQUAL_INT(sizeof(TEST_STRING4), entry->l2_addr_len);
+    TEST_ASSERT_EQUAL_INT(0, entry->flags);
```
Comment (OlegHahm) [inline tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c]: If this the only additional test, why do we need a separate function?

**Comment by authmillenon** [comment]
Rebased

File: tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c (hunk 13 above)
Comment (authmillenon) [inline tests/unittests/tests-ipv6_nc/tests-ipv6_nc.c]: No, this is an interface global search (see https://github.com/RIOT-OS/RIOT/pull/2461/files#diff-bfdde70ccc3110439add7db00ff1edd6R200). But this test is missing in `test_ipv6_nc_add__success();`

**Comment by authmillenon** [comment]
Addressed commends

**Comment by OlegHahm** [comment]
ACK, please squash.

**Comment by authmillenon** [comment]
Rebased

**Comment by authmillenon** [comment]
*Sigh*, after 6 h of idling Travis realized there was some missing documentation. So here we go again -.-

**Comment by authmillenon** [comment]
Finally :D

**Comment by OlegHahm** [comment]
Woohoo!

**Comment by authmillenon** [comment]
(now onwards to #2433?)

**Comment by OlegHahm** [comment]
That's not my assignment. :smirk:
